pkg/client/opensergo_client.go

```diff
@@ -0,0 +1,215 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"github.com/golang/groupcache"
+	"github.com/opensergo/opensergo-go/pkg/common/logging"
+	transportPb "github.com/opensergo/opensergo-go/pkg/proto/transport/v1"
+	"github.com/opensergo/opensergo-go/pkg/transport/subscribe"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"strconv"
+	"sync/atomic"
+	"time"
+)
+
+// OpenSergoClient is the client to communication with opensergo-control-plane.
+type OpenSergoClient struct {
+	transportServiceClient *transportPb.OpenSergoUniversalTransportServiceClient
+	subscribeConfigStream  transportPb.OpenSergoUniversalTransportService_SubscribeConfigClient
+
+	subscribeDataCache            *subscribe.SubscribeDataCache
+	subscriberRegistry            *subscribe.SubscriberRegistry
+	subscribeConfigStreamObserver *SubscribeConfigStreamObserver
+
+	status    *atomic.Value
+	requestId groupcache.AtomicInt
+}
+
+type OpensergoClientStatus uint8
+
+const (
+	initial OpensergoClientStatus = iota
+	starting
+	started
+	interrupted
+)
+
+// NewOpenSergoClient returns an instance of OpenSergoClient, and init some properties.
+func NewOpenSergoClient(host string, port int) *OpenSergoClient {
+	address := host + ":" + strconv.Itoa(port)
+	clientConn, _ := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	transportServiceClient := transportPb.NewOpenSergoUniversalTransportServiceClient(clientConn)
+
+	status := new(atomic.Value)
+	status.Store(initial)
+	subscribeDataCache := new(subscribe.SubscribeDataCache)
+	subscriberRegistry := new(subscribe.SubscriberRegistry)
+
+	return &OpenSergoClient{
+		transportServiceClient: &transportServiceClient,
+		subscribeDataCache:     subscribeDataCache,
+		subscriberRegistry:     subscriberRegistry,
+		status:                 status,
+		requestId:              groupcache.AtomicInt(0),
+	}
+}
+
+func (openSergoClient OpenSergoClient) GetSubscribeDataCache() *subscribe.SubscribeDataCache {
+	return openSergoClient.subscribeDataCache
+}
+
+func (openSergoClient OpenSergoClient) GetSubscriberRegistry() *subscribe.SubscriberRegistry {
+	return openSergoClient.subscriberRegistry
+}
+
+// RegisterSubscribeInfo
+//
+// registry Subscriber
+// if subscribeConfigClient in OpenSergoClient is initialized then subscribe config-data by subscribeKey
+func (openSergoClient *OpenSergoClient) RegisterSubscribeInfo(subscribeInfo *SubscribeInfo) *OpenSergoClient {
```

Comment:
Maybe we could consider *option-style* API like: `SubscribeConfig(target, opts...)`
where opts resembles `WithSubscriber(xxx)`, `WithLabelSelector(xxx)`.

Reply:
i will add it into the issue list